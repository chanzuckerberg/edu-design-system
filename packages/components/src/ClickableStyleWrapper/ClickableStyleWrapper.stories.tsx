import { Story } from "@storybook/react/types-6-0";
import React from "react";
import ClickableStyleWrapper from "./ClickableStyleWrapper";

export default {
  title: "ClickableStyleWrapper",
  component: ClickableStyleWrapper,
};

type Args = React.ComponentProps<typeof ClickableStyleWrapper>;

const Template: Story<Args> = (args) => <ClickableStyleWrapper {...args} />;

export const ClickableStyleWrapperAsAnATag = Template.bind(null);
ClickableStyleWrapperAsAnATag.args = {
  as: "a",
  color: "brand",
  variant: "flat",
  size: "medium",
  children: "ClickableStyleWrapper As An A Tag",
};
