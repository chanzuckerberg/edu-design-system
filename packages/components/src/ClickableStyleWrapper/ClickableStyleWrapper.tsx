import clsx from "clsx";
import React from "react";
import styles from "./ClickableStyleWrapper.module.css";

export type ClickableStyleWrapperProps<IComponent extends React.ElementType> = {
  /**
   * CSS class for custom styles.
   */
  className: string;
  /**
   * The color of the element.
   */
  color: "alert" | "brand" | "neutral" | "success" | "warning";
  /**
   * The size of the element.
   */
  size: "small" | "medium" | "large";
  /**
   * A hidden prop for visual testing
   */
  state?: "inactive" | "hover" | "focus" | "active";
  /**
   * The style of the element.
   */
  variant: "flat" | "outline" | "link";
} & React.ComponentProps<IComponent>;

/**
 * A helper component that contains all the styling for buttons and links.
 *
 * See the Button stories for usage examples.
 */
const ClickableStyleWrapper = React.forwardRef(
  <IComponent extends React.ElementType>(
    {
      as: Component,
      children,
      color,
      size,
      state,
      variant,
      className,
      ...rest
    }: ClickableStyleWrapperProps<IComponent>,
    ref: React.ForwardedRef<HTMLElement>,
  ) => {
    return (
      <Component
        className={clsx(
          className,
          styles.button,
          // Sizes
          variant !== "link" && [
            size === "small" && styles.sizeSmall,
            size === "medium" && styles.sizeMedium,
            size === "large" && styles.sizeLarge,
          ],
          // Variants
          variant === "flat" && styles.variantFlat,
          variant === "outline" && styles.variantOutline,
          variant === "link" && styles.variantLink,
          // Colors
          color === "alert" && styles.colorAlert,
          color === "brand" && styles.colorBrand,
          color === "neutral" && styles.colorNeutral,
          color === "success" && styles.colorSuccess,
          color === "warning" && styles.colorWarning,
          // Interactive States (for testing)
          state === "hover" && styles.stateHover,
          state === "focus" && styles.stateFocus,
          state === "active" && styles.stateActive,
        )}
        ref={ref}
        {...rest}
      >
        {/* No width space to ensure height of contents */}
        {"\u200B"}
        {children}
      </Component>
    );
  },
);

ClickableStyleWrapper.displayName = "ClickableStyleWrapper";

export default ClickableStyleWrapper;
