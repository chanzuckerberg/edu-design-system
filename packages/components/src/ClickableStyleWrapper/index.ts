export { default } from "./ClickableStyleWrapper";
export type { ClickableStyleWrapperProps } from "./ClickableStyleWrapper";
