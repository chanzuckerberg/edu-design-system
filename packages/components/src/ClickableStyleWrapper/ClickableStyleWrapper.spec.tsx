import { generateSnapshots } from "@chanzuckerberg/story-utils";
import * as ClickableStyleWrapperStoryFile from "./ClickableStyleWrapper.stories";

describe("<ClickableStyleWrapper />", () => {
  generateSnapshots(ClickableStyleWrapperStoryFile);
});
