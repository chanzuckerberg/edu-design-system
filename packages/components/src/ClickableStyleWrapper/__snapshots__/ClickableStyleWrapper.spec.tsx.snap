// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ClickableStyleWrapper /> ClickableStyleWrapperAsAnATag story renders snapshot 1`] = `
<a
  class="button sizeMedium variantFlat colorBrand"
>
  â€‹
  ClickableStyleWrapper As An A Tag
</a>
`;
